Data race in universe.union

I observed this data race will running `make test-race` in CI. I've never seen this one before. This is when running the legacy e2e tests. It seems possible this could be an actual product issue.

It passed when I reran, so it only shows up sporadically.
```
==================
WARNING: DATA RACE
Write at 0x00c0025bad10 by goroutine 121:
  github.com/influxdata/flux/stdlib/universe.(*unionTransformation2).Finish()
      /home/builder/project/stdlib/universe/union2.go:100 +0x204
  github.com/influxdata/flux/stdlib/universe.(*unionTransformation2).ProcessMessage()
      /home/builder/project/stdlib/universe/union2.go:44 +0x197
  github.com/influxdata/flux/execute.(*consecutiveTransport).processMessage()
      /home/builder/project/execute/transport.go:263 +0x181
  github.com/influxdata/flux/execute.(*consecutiveTransport).processMessages()
      /home/builder/project/execute/transport.go:218 +0x177
  github.com/influxdata/flux/execute.(*consecutiveTransport).processMessages-fm()
      /home/builder/project/execute/transport.go:213 +0x57
  github.com/influxdata/flux/execute.(*poolDispatcher).doWork()
      /home/builder/project/execute/dispatcher.go:181 +0x30a
  github.com/influxdata/flux/execute.(*poolDispatcher).run()
      /home/builder/project/execute/dispatcher.go:159 +0x69
  github.com/influxdata/flux/execute.(*poolDispatcher).Start.func1()
      /home/builder/project/execute/dispatcher.go:99 +0x111

Previous read at 0x00c0025bad10 by goroutine 83:
  github.com/influxdata/flux/stdlib/universe.(*unionTransformation2).ProcessMessage()
      /home/builder/project/stdlib/universe/union2.go:33 +0xef
  github.com/influxdata/flux/execute.(*consecutiveTransport).processMessage()
      /home/builder/project/execute/transport.go:263 +0x181
  github.com/influxdata/flux/execute.(*consecutiveTransport).processMessages()
      /home/builder/project/execute/transport.go:218 +0x177
  github.com/influxdata/flux/execute.(*consecutiveTransport).processMessages-fm()
      /home/builder/project/execute/transport.go:213 +0x57
  github.com/influxdata/flux/execute.(*poolDispatcher).doWork()
      /home/builder/project/execute/dispatcher.go:181 +0x30a
  github.com/influxdata/flux/execute.(*poolDispatcher).run()
      /home/builder/project/execute/dispatcher.go:159 +0x69
  github.com/influxdata/flux/execute.(*poolDispatcher).Start.func1()
      /home/builder/project/execute/dispatcher.go:99 +0x111

Goroutine 121 (running) created at:
  github.com/influxdata/flux/execute.(*poolDispatcher).Start()
      /home/builder/project/execute/dispatcher.go:76 +0x5a
  github.com/influxdata/flux/execute.(*executionState).do()
      /home/builder/project/execute/executor.go:300 +0x12a
  github.com/influxdata/flux/execute.(*executor).Execute()
      /home/builder/project/execute/executor.go:78 +0x1f3
  github.com/influxdata/flux/lang.(*Program).Start()
      /home/builder/project/lang/compiler.go:306 +0xcc7
  github.com/influxdata/flux/lang.(*AstProgram).Start()
      /home/builder/project/lang/compiler.go:510 +0xb44
  github.com/influxdata/flux/stdlib_test.doTestRun()
      /home/builder/project/stdlib/flux_test.go:134 +0x6f3
  github.com/influxdata/flux/stdlib_test.testFlux()
      /home/builder/project/stdlib/flux_test.go:110 +0x364
  github.com/influxdata/flux/stdlib_test.runEndToEnd.func3.1()
      /home/builder/project/stdlib/flux_test.go:83 +0x1b3
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47

Goroutine 83 (running) created at:
  github.com/influxdata/flux/execute.(*poolDispatcher).Start()
      /home/builder/project/execute/dispatcher.go:76 +0x5a
  github.com/influxdata/flux/execute.(*executionState).do()
      /home/builder/project/execute/executor.go:300 +0x12a
  github.com/influxdata/flux/execute.(*executor).Execute()
      /home/builder/project/execute/executor.go:78 +0x1f3
  github.com/influxdata/flux/lang.(*Program).Start()
      /home/builder/project/lang/compiler.go:306 +0xcc7
  github.com/influxdata/flux/lang.(*AstProgram).Start()
      /home/builder/project/lang/compiler.go:510 +0xb44
  github.com/influxdata/flux/stdlib_test.doTestRun()
      /home/builder/project/stdlib/flux_test.go:134 +0x6f3
  github.com/influxdata/flux/stdlib_test.testFlux()
      /home/builder/project/stdlib/flux_test.go:110 +0x364
  github.com/influxdata/flux/stdlib_test.runEndToEnd.func3.1()
      /home/builder/project/stdlib/flux_test.go:83 +0x1b3
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
==================
--- FAIL: TestFluxEndToEnd (48.33s)
    --- FAIL: TestFluxEndToEnd/union (0.17s)
        flux_test.go:179: Result: want
            Table: keys: [_field, _measurement, cpu, host]
                     _field:string     _measurement:string              cpu:string             host:string                      _time:time                  _value:float  
            ----------------------  ----------------------  ----------------------  ----------------------  ------------------------------  ----------------------------  
                       usage_guest                     cpu               cpu-total              host.local  2018-05-22T19:53:26.000000000Z                             0  
                       usage_guest                     cpu               cpu-total              host.local  2018-05-22T19:53:36.000000000Z                             0  
                       usage_guest                     cpu               cpu-total              host.local  2018-05-22T19:53:46.000000000Z                             0  
                       usage_guest                     cpu               cpu-total              host.local  2018-05-22T19:53:56.000000000Z                             0  
                       usage_guest                     cpu               cpu-total              host.local  2018-05-22T19:54:06.000000000Z                             0  
                       usage_guest                     cpu               cpu-total              host.local  2018-05-22T19:54:16.000000000Z                             0  
            Table: keys: [_field, _measurement, cpu, host]
                     _field:string     _measurement:string              cpu:string             host:string                      _time:time                  _value:float  
            ----------------------  ----------------------  ----------------------  ----------------------  ------------------------------  ----------------------------  
                  usage_guest_nice                     cpu               cpu-total              host.local  2018-05-22T19:53:26.000000000Z                             0  
                  usage_guest_nice                     cpu               cpu-total              host.local  2018-05-22T19:53:36.000000000Z                             0  
                  usage_guest_nice                     cpu               cpu-total              host.local  2018-05-22T19:53:46.000000000Z                             0  
            Table: keys: [_field, _measurement, cpu, host]
                     _field:string     _measurement:string              cpu:string             host:string                      _time:time                  _value:float  
            ----------------------  ----------------------  ----------------------  ----------------------  ------------------------------  ----------------------------  
                        usage_idle                     cpu               cpu-total              host.local  2018-05-22T19:53:56.000000000Z             91.02836436336374  
                        usage_idle                     cpu               cpu-total              host.local  2018-05-22T19:54:06.000000000Z               68.304576144036  
                        usage_idle                     cpu               cpu-total              host.local  2018-05-22T19:54:16.000000000Z             87.88598574821853  
            Result: got
            Table: keys: [_field, _measurement, cpu, host]
                     _field:string     _measurement:string              cpu:string             host:string                      _time:time                  _value:float  
            ----------------------  ----------------------  ----------------------  ----------------------  ------------------------------  ----------------------------  
                       usage_guest                     cpu               cpu-total              host.local  2018-05-22T19:53:26.000000000Z                             0  
                       usage_guest                     cpu               cpu-total              host.local  2018-05-22T19:53:36.000000000Z                             0  
                       usage_guest                     cpu               cpu-total              host.local  2018-05-22T19:53:46.000000000Z                             0  
                       usage_guest                     cpu               cpu-total              host.local  2018-05-22T19:53:56.000000000Z                             0  
                       usage_guest                     cpu               cpu-total              host.local  2018-05-22T19:54:06.000000000Z                             0  
                       usage_guest                     cpu               cpu-total              host.local  2018-05-22T19:54:16.000000000Z                             0  
            Table: keys: [_field, _measurement, cpu, host]
                     _field:string     _measurement:string              cpu:string             host:string                      _time:time                  _value:float  
            ----------------------  ----------------------  ----------------------  ----------------------  ------------------------------  ----------------------------  
                  usage_guest_nice                     cpu               cpu-total              host.local  2018-05-22T19:53:26.000000000Z                             0  
                  usage_guest_nice                     cpu               cpu-total              host.local  2018-05-22T19:53:36.000000000Z                             0  
                  usage_guest_nice                     cpu               cpu-total              host.local  2018-05-22T19:53:46.000000000Z                             0  
            Table: keys: [_field, _measurement, cpu, host]
                     _field:string     _measurement:string              cpu:string             host:string                      _time:time                  _value:float  
            ----------------------  ----------------------  ----------------------  ----------------------  ------------------------------  ----------------------------  
                        usage_idle                     cpu               cpu-total              host.local  2018-05-22T19:53:56.000000000Z             91.02836436336374  
                        usage_idle                     cpu               cpu-total              host.local  2018-05-22T19:54:06.000000000Z               68.304576144036  
                        usage_idle                     cpu               cpu-total              host.local  2018-05-22T19:54:16.000000000Z             87.88598574821853  
            Result: diff
            
        testing.go:1152: race detected during execution of test
    testing.go:1152: race detected during execution of test
FAIL
FAIL	github.com/influxdata/flux/stdlib	49.315s
```

--- a/stdlib/universe/union2.go
+++ b/stdlib/universe/union2.go
@@ -28,16 +28,16 @@ func newUnionTransformation2(id execute.DatasetID, parents []execute.DatasetID, 
 func (u *unionTransformation2) ProcessMessage(m execute.Message) error {
 	defer m.Ack()
 
+	// It is possible to receive messages from different parents concurrently.
+	// Lock this here to prevent a race condition.
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
 	// If this transformation was already marked as finished
 	// from an error, do not accept new messages at all.
 	if u.parents == 0 {
 		return nil
 	}
-
-	// It is possible to receive messages from different parents concurrently.
-	// Lock this here to prevent a race condition.
-	u.mu.Lock()
-	defer u.mu.Unlock()
 
 	switch m := m.(type) {
 	case execute.FinishMsg:

